Implement layouts per workspace

kid opened the pull request:
This is an attempt to resolve #618, this is still WIP, but thought I would open the PR for early feedbacks before proceeding. It's still untested and I'm not sure how it will work with `layout_mode`

kid commented on leftwm-core/src/models/layout_manager.rs:
```diff
@@ -19,40 +19,53 @@ impl Default for LayoutMode {
+                Some(index) => layouts.get(index + 1),
+                None => None,
+            })
+            .unwrap_or_default()
```
Should we return the current layout here if we can't find the next one?

AethanFoot commented on leftwm-core/src/models/layout_manager.rs:
```diff
@@ -19,40 +19,53 @@ impl Default for LayoutMode {
+                Some(index) => layouts.get(index + 1),
+                None => None,
+            })
+            .unwrap_or_default()
```
This is quite hard to digest, something like:
```rs
match workspace.id {
    Some(id) if self.layouts_per_workspace.contains_key(&id) => *self.layouts_per_workspaces.get(&id),
    _ => match self.layouts.iter().position(|&l| l == layout) {
        Some(index) if index == layouts.len() - 1 => *layouts.first(),
        Some(index) => *layouts.get(index + 1),
        None => layout,
    }
}
```
Feels a bit easier to read.

And yes I would return the current layout if we can't find the next.

kid replied:
This does  not quiet work as the second part of the match need to apply to either the layouts from the workspace, or the global one, 

Keeping the same style however, this work and is I think quiet readable:
```rust
        let layouts = workspace
            .id
            .and_then(|id| self.layouts_per_workspaces.get(&id))
            .unwrap_or(&self.layouts);

        let next = match layouts.iter().position(|&x| x == current) {
            Some(index) if index == layouts.len() - 1 => layouts.first(),
            Some(index) => layouts.get(index + 1),
            None => None,
        };

        *next.unwrap_or(&current)
```

I've renamed `layout` to `current` for clarity

AethanFoot commented on leftwm-core/src/models/layout_manager.rs:
```diff
@@ -19,40 +19,53 @@ impl Default for LayoutMode {
+            })
+            .unwrap_or_default()
     }
 
```
**Same here.

AethanFoot commented on leftwm-core/src/models/layout_manager.rs:
```diff
@@ -19,40 +19,53 @@ impl Default for LayoutMode {
-            index = 0;
-        }
-        self.layouts[index as usize]
+    pub fn next_layout(&self, layout: Layout, workspace: &Workspace) -> Layout {
```
Should we just pass the id e.g.:
```rs
pub fn next_layout(...., id: Option<i32>)...
```

kid replied:
we could also pass the workspace itself, since everywhere this is called, the layout comes from the workspace anyway

AethanFoot commented on leftwm-core/src/layouts/mod.rs:
```diff
@@ -54,6 +54,12 @@ impl Default for Layout {
+        &Layout::MainAndVertStack
+    }
+}
+
```
This is kinda redundant. With returning the current layout this won't be needed anyway.

AethanFoot reviewed (commented):
Nice work, just had some niggly little things.

kid commented:
I've been using this locally for the last 3 days, and this seems to work perfectly so far

There is however one edge case which is when the display configuration changes.
Basically if you want to define layouts per screen, you're forced to define workspaces that match your screen configuration, which isn't ideal.

In my case, I have a dual monitor setup, one primary ultra-wide, one vertical on the right, because of this, the primary monitor is at (0, 480), but if I disconnect the second monitor (e.g, to have g-sync working properly...), then the primary is now positioned at (0, 0), and that screw up the layout (in this case, the layout starts a (0, 480))

AethanFoot commented:
Ye we need to improve the adaption to screen changes. I should have time tomorrow to test this then can merge.

AethanFoot commented:
This looks like it is ready, only one thing I found that in a practical case the returning the current layout if we can't find a next/previous leads to the user being stuck on a layout if they soft reload with different layouts, maybe we should return the first in the list for that workspace.

kid commented:
Ha yes, that makes sense, fixed that.

I also fixed another edge case where the layout list is defined, but is empty, in that case we should fallback go the global layout list

AethanFoot commented:
In future what we could do is do per screen configuration. This would allow a screen to to have settings like the current workspace config, but it can be defined by its port or say its orientation (that would be options for all in that orientation). The screen then can be configured with multiple workspaces under if someone is on an ultrawide. This would help as we can automatically find the dimensions if not specified and allows options like this to be applied without the sizing and such being needed. This would also hopefully allow greater options for multi-monitor setups and we could add a separate utility that helps the user setup these options.
